### DISCUSSION TITLE
How to generate a wonderful compile_commands.json for header file in gnu make project?

### DISCUSSION THREAD

**Discussion by zhiqiangzz** [clangd/clangd#1778] [Q&A] [answered]
I recently encountered an issue with the compilation database while debugging my `as` project. 
I build `as` with
``` shell
export CFLAG="-O0 -g"
./configure --target=riscv64-linux-gnu 
bear -- make all-gas
```

When a header file includes a pointer to a struct type without needing to see the definition of the struct itself.
In reality, I don't need to include the header file where this struct is defined. However, this also prevents me from directly navigating to the definition of this struct from the header that includes a pointer to it, and Visual Studio Code generates a red error squiggle, which is quite frustrating.
Such as
![CleanShot 2023-10-09 at 14 48 26@2x](https://github.com/clangd/clangd/assets/91303273/9b6979ca-9a2b-4dd4-b454-fafa28599433)

Or when a static function in the header file uses a certain function without including the header file declaring that function.
Actually, there's no need to include the declaration of this external function in static function defined in the header file. It's sufficient to include the declaration of the external function used within the static function in the C file where the static function is used. However, this also prevents me from directly navigating to the definition of the external function used within the static function.
Such as
![CleanShot 2023-10-09 at 14 56 00@2x](https://github.com/clangd/clangd/assets/91303273/235a4fda-95ab-4f77-afef-498717276208)

Because this is a standard GNU project, it's also challenging for me to manually add all the other header files that might be "dependencies" in each header file to generate property compile_commands.json (which also may destroy the debugging info). When I'm debugging using Visual Studio Code, I encounter a lot of red squiggles and navigation issues, which causes significant inconvenience in my debugging process. Is there a good way to address such situations?

Appreciate it!

**Comment by i-ky** [comment]
I don't think you will be able to generate `compile_commands.json` entries for these particular cases with a general purpose tool like `bear`, because these headers (I mean cgen.h, hash.h, etc.) apparently imply extra rules to use them, like including other headers with forward declarations of types and functions. A general purpose tool will not be able to understand these rules from the captured build commands. My suggestion would be to write your own script, which would extend `compile_commands.json` generated by `bear` with compilation commands for these special headers (cgen.h, hash.h, etc.). In this script you need to explicitly implement these implicit header usage rules (e.g. "using `cgen.h` implies `-include foo.h`").

**Reply by zhiqiangzz** [reply]
I agree with what you're saying. However, this project was cloned from the official GNU repository. If I want to write my own script, do I need to modify its Makefile? I find this to be a somewhat challenging process. I apologize for not being very familiar with the automake and autoconf toolset; I usually work with CMake for development. I just want to do some simple debugging on this project.
Could you please provide examples that I can study to learn how to write my own scripts for modification?

**Reply by i-ky** [reply, marked as answer]
> If I want to write my own script, do I need to modify its Makefile?

My idea was that you don't need to modify any files in the project.

> I usually work with CMake for development.

There is no difference, since you will not be modifying the project. `compile_commands.json` format is the same.

> Could you please provide examples that I can study to learn how to write my own scripts for modification?

I tend to use Python for tasks like this, because [`compile_commands.json` is a simple JSON-based format](https://clang.llvm.org/docs/JSONCompilationDatabase.html) and Python has a [standard library module to work with JSON](https://docs.python.org/3/library/json.html):
```python
#!/usr/bin/env python3

from json import load, dump

with open("compile_commands.json") as f:
    j = load(f)
    j.append({"directory": "...", "command": "...", "file": "cgen.h"})
    # or
    j.append({"directory": "...", "arguments": ["...", "...", "..."], "file": "cgen.h"})
    # ...

with open("compile_commands.json", "w") as f:
    dump(j, f, indent=2)
```

But you can also use other languages or CLI tools to manipulate JSON (e.g. [`jq`](https://jqlang.github.io/jq/)). There may be libraries specifically targeted at editing `compile_commands.json` as well.

**Reply by zhiqiangzz** [reply]
I understand what you mean. Thank you very much for your help

**Comment by i-ky** [comment]
If it was your own project, I would suggest adhering to [Include What You Use principle](https://github.com/include-what-you-use/include-what-you-use/blob/master/docs/WhyIWYU.md) to make cgen.h, hash.h, etc. self-sufficient from compiler's perspective.
